iceoryx_examples/callbacks/README.md

```diff
@@ -1,2 +1,138 @@
+#### int main()
+The subscriber main function starts as usual and after registering the runtime 
+we create the listener which starts a background thread.
+```cpp
+iox::popo::Listener listener;
+```
+
+Because it is fun we also create a heartbeat trigger which will be triggered 
+every 4 seconds so that `heartbeat received` can be printed to the console.
+Furthermore, we have to create two subscriber to receive samples for the two 
+services.
+```cpp
+iox::popo::UserTrigger heartbeat;
+iox::popo::Subscriber<CounterTopic> subscriberLeft({"Radar", "FrontLeft", "Counter"});
+iox::popo::Subscriber<CounterTopic> subscriberRight({"Radar", "FrontRight", "Counter"});
+```
+
+Next thing is a `heartbeatThread` which will trigger our heartbeat trigger every 
+4 seconds.
+```cpp
+std::thread heartbeatThread([&] {
+    while (keepRunning)
+    {
+        heartbeat.trigger();
+        std::this_thread::sleep_for(std::chrono::seconds(4));
+    }
+});
+```
+
+Now that everything is setup we can attach the subscribers to the listener so that
+everytime a new sample (`iox::popo::SubscriberEvent::HAS_DATA`) is received our callback 
+(`onSampleReceivedCallback`) will be called. We also attach 
+our `heartbeat` user trigger to print the hearbeat message to the console via another
+callback (`heartbeatCallback`).
+```cpp
+listener.attachEvent(heartbeat, heartbeatCallback);
+listener.attachEvent(subscriberLeft, 
+      iox::popo::SubscriberEvent::HAS_DATA, onSampleReceivedCallback);
+listener.attachEvent(subscriberRight, 
+      iox::popo::SubscriberEvent::HAS_DATA, onSampleReceivedCallback);
+```
+Since a user trigger has only one event we do not have to specify an event when we attach 
+it to the listener.
+
+The setup is complete but it would terminate right away since we have no blocker which
+waits until SIGINT or SIGTERM was send. In the other examples we hadn't have that problem
+since we pulled all the events in a while true loop but working only with callbacks 
+requires something like our `shutdownSemaphore`, a semaphore on which we wait until 
+the signal callback increments it.
+```cpp
+shutdownSemaphore.wait();
+```
+
+When the `shutdownSemaphore` unblocks we clean up all resources and terminate the process 
+gracefully.
+```cpp
+listener.detachEvent(heartbeat);
+listener.detachEvent(subscriberLeft, iox::popo::SubscriberEvent::HAS_DATA);
+listener.detachEvent(subscriberRight, iox::popo::SubscriberEvent::HAS_DATA);
+
+heartbeatThread.join();
+```
+
+Hint: You do not have to detach an _EventOrigin_ like a subscriber or user trigger 
+before it goes out of scope. This also goes for the _Listener_, the implemented
+RAII based design takes care of the resource cleanup.
+
+#### The Callbacks
+The callbacks must have a signature like `void(PointerToEventOrigin*)`.
+Our `heartbeatCallback` for instance just prints the message `heartbeat received`.
+```cpp
+void heartbeatCallback(iox::popo::UserTrigger*)
+{
+    std::cout << "heartbeat received " << std::endl;
```

Comment:
Where do you reset the user trigger? inside the listener? Can it distinguish the attached elements?

Reply:
The user trigger does not require a reset in the Listener case. The event variable traces by whom it was triggered and inside the EventListener.wait call all events are resetted and returned to the user.